File: packages/otel/src/mapping.ts
/**
 * packages/otel/src/mapping.ts
 *
 * Semantic mapping between AeroGraph event kinds and OTel span metadata.
 *
 * Functions:
 *   getSpanNameForKind(kind)         — AeroGraph kind → OTLP span name
 *   getSpanKindInt(kind)             — AeroGraph kind → OTLP span kind integer (0–5)
 *   buildAttributesFromEvent(event)  — AeroGraph TraceEvent → OTLP attributes array
 *   exportLinksToOtlp(links, traceId) — AeroGraph TraceLink[] → OtlpLink[]
 *
 * Mapping table (from research.md §5.1):
 *   prompt          → gen_ai.chat          CLIENT (3)
 *   response        → gen_ai.response      CLIENT (3)
 *   tool_call       → gen_ai.tool.call     CLIENT (3)
 *   tool_result     → gen_ai.tool.result   INTERNAL (1)
 *   handoff         → gen_ai.agent.handoff INTERNAL (1)
 *   error           → aerograph.error      INTERNAL (1)
 *   note            → aerograph.note       INTERNAL (1)
 *   retriever       → gen_ai.retrieve      CLIENT (3)
 *   checkpoint      → aerograph.checkpoint INTERNAL (1)
 *   state_snapshot  → aerograph.state_snapshot INTERNAL (1)
 *
 * Mirrors mapping.py in Python exactly.
 */

import type { TraceEvent, TraceEventKind, LinkRel } from "@aerograph/contracts";

/** AeroGraph trace link (mirrors traceLinkSchema in contracts) */
interface TraceLink {
  rel: LinkRel;
  spanId: string;
}
import { AEROGRAPH_ATTRS } from "./constants.js";
import type { OtlpAttribute, OtlpLink } from "./otlp-schema.js";

// OTel span kind enum values
export const SPAN_KIND = {
  UNSPECIFIED: 0,
  INTERNAL: 1,
  SERVER: 2,
  CLIENT: 3,
  PRODUCER: 4,
  CONSUMER: 5,
} as const;

// OTel status code enum values
export const STATUS_CODE = {
  UNSET: 0,
  OK: 1,
  ERROR: 2,
} as const;

/**
 * Map an AeroGraph event kind to the OTLP span name.
 */
export function getSpanNameForKind(kind: TraceEventKind): string {
  switch (kind) {
    case "prompt":         return "gen_ai.chat";
    case "response":       return "gen_ai.response";
    case "tool_call":      return "gen_ai.tool.call";
    case "tool_result":    return "gen_ai.tool.result";
    case "handoff":        return "gen_ai.agent.handoff";
    case "error":          return "aerograph.error";
    case "note":           return "aerograph.note";
    case "retriever":      return "gen_ai.retrieve";
    case "checkpoint":     return "aerograph.checkpoint";
    case "state_snapshot": return "aerograph.state_snapshot";
  }
}

/**
 * Map an AeroGraph event kind to the OTLP span kind integer.
 */
export function getSpanKindInt(kind: TraceEventKind): number {
  switch (kind) {
    case "prompt":
    case "response":
    case "tool_call":
    case "retriever":
      return SPAN_KIND.CLIENT;
    case "tool_result":
    case "handoff":
    case "error":
    case "note":
    case "checkpoint":
    case "state_snapshot":
      return SPAN_KIND.INTERNAL;
  }
}

/**
 * Build an OTLP attribute with a string value.
 */
function strAttr(key: string, value: string): OtlpAttribute {
  return { key, value: { stringValue: value } };
}

/**
 * Build an OTLP attribute with an integer value.
 */
function intAttr(key: string, value: number): OtlpAttribute {
  return { key, value: { intValue: value } };
}

/**
 * Build an OTLP attribute with a double value.
 */
function doubleAttr(key: string, value: number): OtlpAttribute {
  return { key, value: { doubleValue: value } };
}

/**
 * Build the OTLP attributes array for a given AeroGraph TraceEvent.
 * Includes all universal attributes + kind-specific payload attributes.
 *
 * Attribute order is deterministic: universal first, kind-specific second.
 */
export function buildAttributesFromEvent(event: TraceEvent): OtlpAttribute[] {
  const attrs: OtlpAttribute[] = [];

  // Universal attributes — on every span
  attrs.push(strAttr(AEROGRAPH_ATTRS.SCHEMA_VERSION, event.schemaVersion));
  attrs.push(strAttr(AEROGRAPH_ATTRS.KIND, event.kind));
  attrs.push(strAttr(AEROGRAPH_ATTRS.ACTOR_ID, event.actor.id));
  attrs.push(strAttr(AEROGRAPH_ATTRS.ACTOR_KIND, event.actor.kind));
  if (event.actor.name) {
    attrs.push(strAttr(AEROGRAPH_ATTRS.ACTOR_NAME, event.actor.name));
  }
  attrs.push(strAttr(AEROGRAPH_ATTRS.STATUS, event.status));
  if (event.title) {
    attrs.push(strAttr(AEROGRAPH_ATTRS.TITLE, event.title));
  }

  // Kind-specific payload attributes
  switch (event.kind) {
    case "prompt":
      attrs.push(strAttr(AEROGRAPH_ATTRS.PROMPT_TEXT, event.payload.text));
      attrs.push(strAttr("gen_ai.operation.name", "chat"));
      break;

    case "response":
      attrs.push(strAttr(AEROGRAPH_ATTRS.RESPONSE_TEXT, event.payload.text));
      if (event.payload.streamingTelemetry) {
        const st = event.payload.streamingTelemetry;
        attrs.push(doubleAttr(AEROGRAPH_ATTRS.RESPONSE_TIME_TO_FIRST_TOKEN_MS, st.timeToFirstTokenMs));
        attrs.push(doubleAttr(AEROGRAPH_ATTRS.RESPONSE_TOTAL_DURATION_MS, st.totalDurationMs));
        attrs.push(doubleAttr(AEROGRAPH_ATTRS.RESPONSE_TOKENS_PER_SECOND, st.tokensPerSecond));
        attrs.push(intAttr(AEROGRAPH_ATTRS.RESPONSE_TOKEN_COUNT, st.tokenCount));
      }
      attrs.push(strAttr("gen_ai.operation.name", "chat"));
      break;

    case "tool_call":
      attrs.push(strAttr(AEROGRAPH_ATTRS.TOOL_CALL_INPUT, JSON.stringify(event.payload.input)));
      attrs.push(strAttr("gen_ai.tool.name", event.actor.id));
      break;

    case "tool_result":
      attrs.push(strAttr(AEROGRAPH_ATTRS.TOOL_RESULT_OUTPUT, JSON.stringify(event.payload.output)));
      attrs.push(strAttr("gen_ai.tool.name", event.actor.id));
      break;

    case "handoff":
      attrs.push(strAttr(AEROGRAPH_ATTRS.HANDOFF_FROM_AGENT_ID, event.payload.fromAgentId));
      attrs.push(strAttr(AEROGRAPH_ATTRS.HANDOFF_TO_AGENT_ID, event.payload.toAgentId));
      if (event.payload.reason) {
        attrs.push(strAttr(AEROGRAPH_ATTRS.HANDOFF_REASON, event.payload.reason));
      }
      attrs.push(strAttr("gen_ai.agent.name", event.payload.toAgentId));
      break;

    case "error":
      attrs.push(strAttr(AEROGRAPH_ATTRS.ERROR_MESSAGE, event.payload.message));
      attrs.push(strAttr(AEROGRAPH_ATTRS.ERROR_DETAILS, JSON.stringify(event.payload.details)));
      attrs.push(strAttr("error.type", "aerograph.error"));
      break;

    case "note":
      attrs.push(strAttr(AEROGRAPH_ATTRS.NOTE_PAYLOAD, JSON.stringify(event.payload)));
      break;

    case "retriever":
      attrs.push(strAttr(AEROGRAPH_ATTRS.RETRIEVER_QUERY, event.payload.query));
      attrs.push(intAttr(AEROGRAPH_ATTRS.RETRIEVER_DOCUMENT_COUNT, event.payload.documents.length));
      attrs.push(strAttr("gen_ai.operation.name", "retrieve"));
      break;

    case "checkpoint":
      attrs.push(strAttr(AEROGRAPH_ATTRS.CHECKPOINT_ID, event.payload.checkpointId));
      attrs.push(strAttr(AEROGRAPH_ATTRS.CHECKPOINT_REASON, event.payload.reason));
      break;

    case "state_snapshot":
      attrs.push(strAttr(AEROGRAPH_ATTRS.STATE_SNAPSHOT_NODE_NAME, event.payload.nodeName));
      attrs.push(strAttr(AEROGRAPH_ATTRS.STATE_SNAPSHOT_STATE_HASH, event.payload.stateHash));
      break;
  }

  return attrs;
}

/**
 * Convert AeroGraph TraceLink[] to OTLP OtlpLink[].
 * Each link carries the aerograph.link.rel attribute.
 *
 * @param links - AeroGraph trace links
 * @param traceId - the traceId to use for the link (same trace)
 */
export function exportLinksToOtlp(links: TraceLink[], traceId: string): OtlpLink[] {
  return links.map((link) => ({
    traceId,
    spanId: link.spanId,
    attributes: [
      strAttr(AEROGRAPH_ATTRS.LINK_REL, link.rel),
    ],
  }));
}

File: python/aerograph-otel/src/aerograph_otel/mapping.py
"""
python/aerograph-otel/src/aerograph_otel/mapping.py

Semantic mapping between AeroGraph event kinds and OTel span metadata.
Mirrors mapping.ts in TypeScript exactly.

Functions:
    get_span_name_for_kind(kind)          — AeroGraph kind → OTLP span name
    get_span_kind_for_kind(kind)          — AeroGraph kind → OTLP span kind integer
    build_attributes_from_event(event)    — AeroGraph event dict → OTLP attributes list
    export_links_to_otlp(links, trace_id) — AeroGraph links → OTel links list
"""

from __future__ import annotations

import json
from typing import Any

from aerograph_otel.constants import AeroGraphAttrs

# OTel span kind enum values
SPAN_KIND_UNSPECIFIED = 0
SPAN_KIND_INTERNAL = 1
SPAN_KIND_SERVER = 2
SPAN_KIND_CLIENT = 3
SPAN_KIND_PRODUCER = 4
SPAN_KIND_CONSUMER = 5

# OTel status code enum values
STATUS_CODE_UNSET = 0
STATUS_CODE_OK = 1
STATUS_CODE_ERROR = 2

# Mapping: AeroGraph kind → OTLP span name
_KIND_TO_SPAN_NAME: dict[str, str] = {
    "prompt":         "gen_ai.chat",
    "response":       "gen_ai.response",
    "tool_call":      "gen_ai.tool.call",
    "tool_result":    "gen_ai.tool.result",
    "handoff":        "gen_ai.agent.handoff",
    "error":          "aerograph.error",
    "note":           "aerograph.note",
    "retriever":      "gen_ai.retrieve",
    "checkpoint":     "aerograph.checkpoint",
    "state_snapshot": "aerograph.state_snapshot",
}

# Mapping: AeroGraph kind → OTLP span kind integer
_KIND_TO_SPAN_KIND: dict[str, int] = {
    "prompt":         SPAN_KIND_CLIENT,
    "response":       SPAN_KIND_CLIENT,
    "tool_call":      SPAN_KIND_CLIENT,
    "tool_result":    SPAN_KIND_INTERNAL,
    "handoff":        SPAN_KIND_INTERNAL,
    "error":          SPAN_KIND_INTERNAL,
    "note":           SPAN_KIND_INTERNAL,
    "retriever":      SPAN_KIND_CLIENT,
    "checkpoint":     SPAN_KIND_INTERNAL,
    "state_snapshot": SPAN_KIND_INTERNAL,
}


def get_span_name_for_kind(kind: str) -> str:
    """Map an AeroGraph event kind to the OTLP span name."""
    try:
        return _KIND_TO_SPAN_NAME[kind]
    except KeyError:
        raise ValueError(f"Unknown AeroGraph event kind: {kind!r}")


def get_span_kind_for_kind(kind: str) -> int:
    """Map an AeroGraph event kind to the OTLP span kind integer."""
    try:
        return _KIND_TO_SPAN_KIND[kind]
    except KeyError:
        raise ValueError(f"Unknown AeroGraph event kind: {kind!r}")


def _str_attr(key: str, value: str) -> dict[str, Any]:
    return {"key": key, "value": {"stringValue": value}}


def _int_attr(key: str, value: int) -> dict[str, Any]:
    return {"key": key, "value": {"intValue": value}}


def _double_attr(key: str, value: float) -> dict[str, Any]:
    return {"key": key, "value": {"doubleValue": value}}


def build_attributes_from_event(event: dict[str, Any]) -> list[dict[str, Any]]:
    """
    Build the OTLP attributes list for a given AeroGraph TraceEvent dict.

    Attribute order is deterministic: universal first, kind-specific second.
    Mirrors buildAttributesFromEvent in TypeScript exactly.
    """
    attrs: list[dict[str, Any]] = []
    kind = event["kind"]
    actor = event["actor"]
    payload = event.get("payload", {})

    # Universal attributes
    attrs.append(_str_attr(AeroGraphAttrs.SCHEMA_VERSION, event["schemaVersion"]))
    attrs.append(_str_attr(AeroGraphAttrs.KIND, kind))
    attrs.append(_str_attr(AeroGraphAttrs.ACTOR_ID, actor["id"]))
    attrs.append(_str_attr(AeroGraphAttrs.ACTOR_KIND, actor["kind"]))
    if actor.get("name"):
        attrs.append(_str_attr(AeroGraphAttrs.ACTOR_NAME, actor["name"]))
    attrs.append(_str_attr(AeroGraphAttrs.STATUS, event["status"]))
    if event.get("title"):
        attrs.append(_str_attr(AeroGraphAttrs.TITLE, event["title"]))

    # Kind-specific payload attributes
    if kind == "prompt":
        attrs.append(_str_attr(AeroGraphAttrs.PROMPT_TEXT, payload["text"]))
        attrs.append(_str_attr("gen_ai.operation.name", "chat"))

    elif kind == "response":
        attrs.append(_str_attr(AeroGraphAttrs.RESPONSE_TEXT, payload["text"]))
        st = payload.get("streamingTelemetry")
        if st:
            attrs.append(_double_attr(AeroGraphAttrs.RESPONSE_TIME_TO_FIRST_TOKEN_MS, st["timeToFirstTokenMs"]))
            attrs.append(_double_attr(AeroGraphAttrs.RESPONSE_TOTAL_DURATION_MS, st["totalDurationMs"]))
            attrs.append(_double_attr(AeroGraphAttrs.RESPONSE_TOKENS_PER_SECOND, st["tokensPerSecond"]))
            attrs.append(_int_attr(AeroGraphAttrs.RESPONSE_TOKEN_COUNT, st["tokenCount"]))
        attrs.append(_str_attr("gen_ai.operation.name", "chat"))

    elif kind == "tool_call":
        attrs.append(_str_attr(AeroGraphAttrs.TOOL_CALL_INPUT, json.dumps(payload["input"], separators=(",", ":"))))
        attrs.append(_str_attr("gen_ai.tool.name", actor["id"]))

    elif kind == "tool_result":
        attrs.append(_str_attr(AeroGraphAttrs.TOOL_RESULT_OUTPUT, json.dumps(payload["output"], separators=(",", ":"))))
        attrs.append(_str_attr("gen_ai.tool.name", actor["id"]))

    elif kind == "handoff":
        attrs.append(_str_attr(AeroGraphAttrs.HANDOFF_FROM_AGENT_ID, payload["fromAgentId"]))
        attrs.append(_str_attr(AeroGraphAttrs.HANDOFF_TO_AGENT_ID, payload["toAgentId"]))
        if payload.get("reason"):
            attrs.append(_str_attr(AeroGraphAttrs.HANDOFF_REASON, payload["reason"]))
        attrs.append(_str_attr("gen_ai.agent.name", payload["toAgentId"]))

    elif kind == "error":
        attrs.append(_str_attr(AeroGraphAttrs.ERROR_MESSAGE, payload["message"]))
        attrs.append(_str_attr(AeroGraphAttrs.ERROR_DETAILS, json.dumps(payload.get("details", {}), separators=(",", ":"))))
        attrs.append(_str_attr("error.type", "aerograph.error"))

    elif kind == "note":
        attrs.append(_str_attr(AeroGraphAttrs.NOTE_PAYLOAD, json.dumps(payload, separators=(",", ":"))))

    elif kind == "retriever":
        attrs.append(_str_attr(AeroGraphAttrs.RETRIEVER_QUERY, payload["query"]))
        attrs.append(_int_attr(AeroGraphAttrs.RETRIEVER_DOCUMENT_COUNT, len(payload["documents"])))
        attrs.append(_str_attr("gen_ai.operation.name", "retrieve"))

    elif kind == "checkpoint":
        attrs.append(_str_attr(AeroGraphAttrs.CHECKPOINT_ID, payload["checkpointId"]))
        attrs.append(_str_attr(AeroGraphAttrs.CHECKPOINT_REASON, payload["reason"]))

    elif kind == "state_snapshot":
        attrs.append(_str_attr(AeroGraphAttrs.STATE_SNAPSHOT_NODE_NAME, payload["nodeName"]))
        attrs.append(_str_attr(AeroGraphAttrs.STATE_SNAPSHOT_STATE_HASH, payload["stateHash"]))

    return attrs


def export_links_to_otlp(links: list[dict[str, Any]], trace_id: str) -> list[dict[str, Any]]:
    """
    Convert AeroGraph TraceLink list to OTLP links list.
    Each link carries the aerograph.link.rel attribute.

    Mirrors exportLinksToOtlp in TypeScript exactly.
    """
    return [
        {
            "traceId": trace_id,
            "spanId": link["spanId"],
            "attributes": [
                _str_attr(AeroGraphAttrs.LINK_REL, link["rel"]),
            ],
        }
        for link in links
    ]
